File: DSA/Linked list/Rev-Rotat-Merge/MergeTwoLInkedlist.java


public class MergeTwoLInkedlist {
    // Node class
    static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    // Merge two sorted linked lists
    public static Node mergeTwoLists(Node l1, Node l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;

        if (l1.data < l2.data) {
            l1.next = mergeTwoLists(l1.next, l2);
            return l1;
        } else {
            l2.next = mergeTwoLists(l1, l2.next);
            return l2;
        }
    }

    // Display the linked list
    public static void display(Node head) {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Create first linked list: 1 -> 3 -> 5
        Node l1 = new Node(1);
        l1.next = new Node(3);
        l1.next.next = new Node(5);

        // Create second linked list: 2 -> 4 -> 6
        Node l2 = new Node(2);
        l2.next = new Node(4);
        l2.next.next = new Node(6);

        // Merge the two linked lists
        Node mergedList = mergeTwoLists(l1, l2);

        // Display the merged linked list
        display(mergedList);
    }
    
}

File: DSA/Linked list/Rev-Rotat-Merge/MergeTwoLinkedlist.py
# merge two linked list
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
class LinkedList:
    def __init__(self):
        self.head = None

    def push(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head
        self.head = new_node

    def merge(self, list1, list2):
        if not list1:
            return list2
        if not list2:
            return list1
        if list1.data < list2.data:
            list1.next = self.merge(list1.next, list2)
            return list1
        else:
            list2.next = self.merge(list1, list2.next)
            return list2
    def printList(self):
        temp = self.head
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print()
# Driver code
if __name__ == "__main__":
    list1 = LinkedList()
    list2 = LinkedList()
    list1.push(15)
    list1.push(10)
    list1.push(5)
    list2.push(20)
    list2.push(3)
    list2.push(2)

    print("First Linked List:")
    list1.printList()

    print("Second Linked List:")
    list2.printList()

    merged_list = LinkedList()
    merged_list.head = merged_list.merge(list1.head, list2.head)

    print("Merged Linked List:")
    merged_list.printList()